File: Apps/BankBase/backend/src/main/scala/security/SecurityService.scala
package security

import errors.AuthError
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import zio.{RLayer, Task, ZIO, ZLayer}

case class SecurityService(private val passwordEncoder: BCryptPasswordEncoder) {

  def hashPassword(password: String): Task[String] = {
    ZIO.succeed(passwordEncoder.encode(password))
  }

  def verifyPassword(password: String, hashedPassword: String): Task[Boolean] = {
    ZIO.succeed(passwordEncoder.matches(password, hashedPassword))
      .flatMap({
        case true => ZIO.succeed(true)
        case false => ZIO.fail(AuthError("Wrong password"))
      })
  }
}

object SecurityService {
  val layer: RLayer[BCryptPasswordEncoder, SecurityService] = ZLayer.fromFunction(SecurityService.apply _)
}

File: Apps/AuthMS/service/security/security_service.go
package security

import (
	"log/slog"

	"golang.org/x/crypto/bcrypt"
)

type SecurityService interface {
	HashPassword(password string) (*string, error)
	VerifyPassword(password string, hash string) error
}

type SecurityServiceImpl struct{}

func (_ *SecurityServiceImpl) HashPassword(password string) (*string, error) {
	if hash, err := bcrypt.GenerateFromPassword([]byte(password), 13); err != nil {
		slog.Error("Password hashing error: ", err)
		return nil, err
	} else {
		hashed := string(hash)
		return &hashed, nil
	}
}

func (_ *SecurityServiceImpl) VerifyPassword(password string, hash string) error {
	return bcrypt.CompareHashAndPassword([]byte(hash), []byte(password))
}
